### PR TITLE
Add -start-vpnkit flag to linuxkit run hyperkit

### PR DESCRIPTION
**Pull request by MagnusS**
Adds a flag to launch a new vpnkit instance. This enables port
forwarding to be configured via a 9p mount (tag:port) if the guest 
supports it. Vsock port 62373 is also opened so vpnkit can forward 
connections back to the guest.

If -start-vpnkit is not specified HyperKit will connect to the
vpnkit instance managed by Docker for Mac, as before.

### CODE REVIEW

**Comment by MagnusS** [comment]
This also enables network access from the VM without a running DfM

File: src/cmd/linuxkit/run_hyperkit.go [hunk 1]
```diff
@@ -6,13 +6,75 @@ import (
...
 	"os"
+	"os/exec"
 	"path/filepath"
 
 	log "github.com/Sirupsen/logrus"
 	"github.com/moby/hyperkit/go"
 	"github.com/satori/go.uuid"
 )
 
+var defaultVPNKits = []string{"vpnkit",
```
Comment (riyazdf) [inline src/cmd/linuxkit/run_hyperkit.go]: nit: can we newline for the individual entries and closing brace?  It'll make for cleaner diffs if we ever edit this

File: src/cmd/linuxkit/run_hyperkit.go [hunk 2]
```diff
@@ -6,13 +6,75 @@ import (
...
+	"/usr/local/bin/vpnkit",
+	"/Applications/Docker.app/Contents/Resources/bin/hyperkit"}
+
+func createListenSocket(path string) (*os.File, error) {
```
Comment (rneugeba) [inline src/cmd/linuxkit/run_hyperkit.go]: could we move the vpnkit functions further to the end. I seem to recall that in go you are supposed to have the exported/main functions first and have the internal ones later in the file. To me as a C programmer that's weird, but hey...I certainly wouldn't put them at the start as the main purpose of this file to run hyperkit

File: src/cmd/linuxkit/run_hyperkit.go [hunk 3]
```diff
@@ -6,13 +6,75 @@ import (
...
+	"os/exec"
 	"path/filepath"
 
 	log "github.com/Sirupsen/logrus"
 	"github.com/moby/hyperkit/go"
 	"github.com/satori/go.uuid"
 )
 
+var defaultVPNKits = []string{"vpnkit",
+	"/usr/local/bin/vpnkit",
```
Comment (justincormack) [inline src/cmd/linuxkit/run_hyperkit.go]: is this needed? lookpath will find it anyway from "vpnkit"

File: src/cmd/linuxkit/run_hyperkit.go [hunk 4]
```diff
@@ -116,6 +202,12 @@ func runHyperKit(args []string) {
 	h.Memory = *mem
 	h.DiskSize = *diskSz
 
+	// Add 9p and vsock for port forwarding if VPNKit is launched automatically
+	if *startVPNKit {
+		h.Sockets9P = []hyperkit.Socket9P{{Path: vpnKitPortSocket, Tag: "port"}}
+		h.VSockPorts = append(h.VSockPorts, 62373) // forwarding
```
Comment (justincormack) [inline src/cmd/linuxkit/run_hyperkit.go]: whats `62373`? Could do with a comment as it is non obvious.

**Comment by MagnusS** [comment]
@rneugeba Yes, if you want to have multiple VMs in the same network you would need an option for specifying the vpnkit socket. It could perhaps be useful if we want to do automated testing of network services with several nodes without having to install full DfM (you would only need hyperkit + vpnkit binaries). But it should be easy to add support for that later if we need it by adding an option to specify a custom vpnkit socket instead of using the DfM one by default. 

Only the node that was started with `-start-vpnkit` would be able to forward ports from the host though.

File: src/cmd/linuxkit/run_hyperkit.go (hunk 3 above)
Comment (MagnusS) [inline src/cmd/linuxkit/run_hyperkit.go]: Hm no, I'll remove it - if you have a relatively new DfM installed it should be in path, and older versions wouldn't have the features we need.

File: src/cmd/linuxkit/run_hyperkit.go (hunk 4 above)
Comment (MagnusS) [inline src/cmd/linuxkit/run_hyperkit.go]: It's the vsock port used to forward network traffic from host->guest. I'll add a comment

File: src/cmd/linuxkit/run_hyperkit.go (hunk 1 above)
Comment (MagnusS) [inline src/cmd/linuxkit/run_hyperkit.go]: Good point, I'll do that in the future

File: src/cmd/linuxkit/run_hyperkit.go (hunk 2 above)
Comment (MagnusS) [inline src/cmd/linuxkit/run_hyperkit.go]: Yes, I'll move them to the end - not used to that either

**Comment by MagnusS** [comment]
Thanks for the reviews. Force pushed a new version with the following changes:
- Only look for "vpnkit" in path
- Move vpnkit helpers to the end
- More descriptive comments on the automatically added 9p mount/vsock port and the helper functions

**Comment by justincormack** [comment]
@MagnusS there is interest in having local swarm clusters so I think adding an option to specify the socket would be useful (and we should stop relying on d4m as we wont have that in CI)

**Pull request merged**